lightningd/peer_control.c

```diff
@@ -562,6 +562,83 @@ struct peer *peer_by_id(struct lightningd *ld, const struct pubkey *id)
 	return NULL;
 }
 
+/* If the given 'id' is of the form xxxxx@yyy:zzz, then
+ * split it into id=xxxxx, host=yyy, port=zzz.
+ * Or, if the given 'id is of the form xxxxx@yyy, then
+ * split it into id=xxxxx, host=yyy
+ *
+ * Return true if no problems encountered (incuding not
+ * detecting xxxxx@yyy:zzz or xxxxx@yyy form), false if
+ * we errored and caller should return to dispatcher */
+static bool connect_standard_node_address(struct command *cmd,
+					  const char* buffer,
+					  jsmntok_t **id,
+					  jsmntok_t **host,
+					  jsmntok_t **port)
+{
+	/* Indexes at which we found @ and : */
+	int found_at = -1;
+	int found_colon = -1;
+	int i;
+	int orig_end = (*id)->end;
+
+	assert(*id);
+
+	for (i = (*id)->start; i < (*id)->end; ++i) {
+		if (found_at == -1) {
+			if (buffer[i] == '@')
+				found_at = i;
+		} else {
+			if (buffer[i] == ':') {
+				found_colon = i;
```

Comment:
This won't work for IPv6 addresses, unfortunately.

You need to expose `separate_address_and_port` and use it here, I think.  It's currently buried in common/wireaddr.c.

And I think that means you'll want to use strchr to find the first `@` rather than open-coding the loop, once that logic is removed.

Reply:
Why not use `parse_wireaddr` instead? We need to build a `struct wireaddr` anyway.  We just have two main cases --- one where the `@` exists, one where `@` is not exist. Without `@` we use current processing.  With `@`, default `port` passed to `parse_wireaddr` is either `DEFAULT_PORT` (if `!porttok`) or `json_tok_number` of the `porttok`.